spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/pubsub/brave/PubSubTracing.java

```diff
@@ -0,0 +1,316 @@
+	}
+
+	/** Creates an instrumented {@linkplain SubscriberStub} for use in message pulling scenario. */
+	public TracingSubscriberStub subscriberStub(SubscriberStub subscriberStub) {
+		return new TracingSubscriberStub(subscriberStub, this);
+	}
+
+	/** Creates an instrumented {@linkplain MessageReceiver} for use in message listening scenario. */
+	public TracingMessageReceiver messageReceiver(MessageReceiver messageReceiver, String subscriptionName) {
+		return new TracingMessageReceiver(messageReceiver, this, subscriptionName);
+	}
+
+	/**
+	 * Use this to create a span for processing the given message. Note: the result has no name and is
+	 * not started.
+	 *
+	 * <p>This creates a child from identifiers extracted from the message headers, or a new span if
+	 * one couldn't be extracted.
+	 */
+	public Span nextSpan(PubsubMessage.Builder message) {
+		// Even though the type is PubSubMessage.Builder, this is not a (remote) consumer span. Only "pull/subscribe"
+		// events create consumer spans. Since this is a processor span, we use the normal sampler.
+		TraceContextOrSamplingFlags extracted =
+				extractAndClearTraceIdHeaders(processorExtractor, message, message);
+		Span result = tracer.nextSpan(extracted);
+		if (extracted.context() == null && !result.isNoop()) {
+			addTags(message, result);
+		}
+		return result;
+	}
+
+	/** Creates a potentially noop remote span representing this request. */
+	Span nextMessagingSpan(
+			SamplerFunction<MessagingRequest> sampler,
+			MessagingRequest request,
+			TraceContextOrSamplingFlags extracted
+	) {
+		Boolean sampled = extracted.sampled();
+		// only recreate the context if the messaging sampler made a decision
+		if (sampled == null) {
+			sampled = sampler.trySample(request);
+			if (sampled != null) {
+				extracted = extracted.sampled(sampled.booleanValue());
+			}
+		}
+		return tracer.nextSpan(extracted);
+	}
+
+	<R> TraceContextOrSamplingFlags extractAndClearTraceIdHeaders(
+			Extractor<R> extractor, R request, PubsubMessage.Builder message
+	) {
+		TraceContextOrSamplingFlags extracted = extractor.extract(request);
+		// Clear any propagation keys present in the headers
+		if (extracted.samplingFlags() == null) { // then trace IDs were extracted
+			clearTraceIdHeaders(message);
+		}
+		return extracted;
+	}
+
+	// We can't just skip clearing headers we use because we might inject B3 single, yet have stale B3
+	// multi, or visa versa.
+	void clearTraceIdHeaders(PubsubMessage.Builder message) {
+		for (String traceIDHeader : traceIdHeaders) {
+			message.removeAttributes(traceIDHeader);
+		}
+	}
+
+	/** When an upstream context was not present, lookup keys are unlikely added */
+	static void addTags(PubsubMessage.Builder message, SpanCustomizer result) {
+
+		// TODO: anything to do here?
+	}
+
+
+	PullResponse tracePullResponse(PullResponse delegate, String subscriptionName) {
+
+		if (delegate.getReceivedMessagesCount() == 0 || tracing.isNoop()) {
```

Comment:
Could "no messages received" be a useful trace signal, too?

Reply:
It's a good question. I don't know. Maybe it would add too much noise. I haven't seen other implementation do anything with an empty response. So, I was just following that pattern established in `KafkaTracing` and others.